### PR TITLE
stats: Reindex documents after the statistics have been calculated

### PR DESCRIPTION
**Pull request by psaiz**
:heart: Thank you for your contribution!

### Description

Addresses https://github.com/zenodo/rdm-project/issues/203

List of changes:
* Retrieves the statistics that have already been generated, and skips entries that are not modified
* Puts the parents of the entries that are updated in a new index as a buffer
* Consumes the entries from that buffer and does a Reindex, `current_rdm_records_service.reindex_latest_first` (THIS PART DOES NOT WORK YET!)
### Checklist

Ticks in all boxes and 🟢 on all GitHub actions status checks are required to merge:

- [x] I'm aware of the [code of conduct](https://inveniordm.docs.cern.ch/contribute/code-of-conduct/).
- [ ] I've created [logical separate commits](https://inveniordm.docs.cern.ch/develop/best-practices/commits/#commits) and followed the [commit message format](https://inveniordm.docs.cern.ch/develop/best-practices/commits/#commit-message).
- [ ] I've added relevant test cases.
- [ ] I've added relevant documentation.
- [x] I've marked [translation strings](https://inveniordm.docs.cern.ch/develop/best-practices/i18n/) (for relevant code).
- [x] I've followed the [CSS/JS](https://inveniordm.docs.cern.ch/develop/best-practices/css-js/) and [React](https://inveniordm.docs.cern.ch/develop/best-practices/react/) guidelines (for relevant code).
- [x] I've followed the [web accessibility](https://inveniordm.docs.cern.ch/develop/best-practices/accessibility/) guidelines (for relevant code).
- [x] I've followed the [user interface](https://inveniordm.docs.cern.ch/develop/best-practices/ui/) guidelines (for relevant code).
- [x] I've identified the [copyright holder(s)](https://inveniordm.docs.cern.ch/contribute/copyright-policy/) and updated copyright headers for touched files (>15 lines contributions).
- [x] I've NOT included third-party code (copy/pasted source code or new dependencies).

**Third-party code**

If you've added [third-party code (copy/pasted or new dependencies)](https://inveniordm.docs.cern.ch/develop/best-practices/commits/#third-party-codedependencies), please reach out to an [architect](https://github.com/orgs/inveniosoftware/teams/architects/members).

**Reminder**

By using GitHub, you have already agreed to the [GitHub’s Terms of Service](https://help.github.com/articles/github-terms-of-service/#6-contributions-under-repository-license) including that:

1. You license your contribution under the same terms as the current repository’s license.
2. You agree that you have the right to license your contribution under the current repository’s license.

### CODE REVIEW

File: invenio_stats/aggregations.py [hunk 1]
```diff
@@ -292,19 +294,37 @@ def _split_date_range(self, lower_limit, upper_limit):
...
 
+    def _get_stats_done(self, index_name, timestamp):
+        """ Retrieve the statistics that have already been inserted """
+        my_query = dsl.Search(
+            using=self.client,
+            index=index_name,
+        ).filter("term", timestamp=timestamp,).\
+            extra(size=self.max_bucket_size,).\
+            source(includes=['count']).execute()
+        print(my_query)
```
Comment (psaiz) [inline invenio_stats/aggregations.py:305]: debug message... will be removed

File: invenio_stats/aggregations.py [hunk 2]
```diff
@@ -292,19 +294,37 @@ def _split_date_range(self, lower_limit, upper_limit):
...
+        # Putting them into a hash for faster checking
+        already_done_hash = {}
+        for entry in my_query:
+            # Let's remove the date from the id
+            already_done_hash[entry.meta['id'][:-11]] = entry.get('count')
+        print("ENTRIES DONE")
```
Comment (psaiz) [inline invenio_stats/aggregations.py:311]: debug

File: invenio_stats/aggregations.py [hunk 3]
```diff
@@ -360,18 +387,46 @@ def agg_iter(self, dt):
                         aggregation_data[destination] = doc[source]
                     else:
                         aggregation_data[destination] = source(doc, aggregation_data)
-
-                index_name = "stats-{0}-{1}".format(
-                    self.event, interval_date.strftime(self.index_name_suffix)
-                )
-
+                print(aggregation_data)
```
Comment (psaiz) [inline invenio_stats/aggregations.py:390]: debug

File: invenio_stats/aggregations.py [hunk 4]
```diff
@@ -360,18 +387,46 @@ def agg_iter(self, dt):
...
+            using=self.client,
+            index=buffer_index,
+        ).source(includes=['timestamp']).execute()
+        latest=None
+        all_parents = []
+        for doc in documents:
+            all_parents.append(doc.meta['id'])
+            if not latest or latest < doc.get('timestamp'):
+                latest=doc.get('timestamp')
+        records_q = dsl.Q("terms", parent_recid=all_parents)
```
Comment (ntarocco) [inline invenio_stats/aggregations.py:422]:
```suggestion
        records_q = dsl.Q("terms", parent__id=all_parents)
```

File: invenio_stats/aggregations.py [hunk 5]
```diff
@@ -19,6 +19,8 @@
 from invenio_search import current_search_client
 from invenio_search.engine import dsl, search
 from invenio_search.utils import prefix_index
+from invenio_access.permissions import system_identity
+from invenio_rdm_records.proxies import current_rdm_records_service
```
Comment (psaiz) [inline invenio_stats/aggregations.py:23]: This one should not be there

File: invenio_stats/aggregations.py [hunk 6]
```diff
@@ -360,18 +385,49 @@ def agg_iter(self, dt):
...
                     "_source": aggregation_data,
                 }
+                yield {
+                    '_index': prefix_index("buffer_stats_to_reindex"),
+                    '_id': doc["parent_recid"],
+                    '_source': {
+                        "timestamp":  datetime.utcnow(),
+                    },
+                }
+        #THIS CALL (AND THE FOLLOWING METHOD) SHOULD BE MOVED SOMEWHERE ELSE
```
Comment (psaiz) [inline invenio_stats/aggregations.py:402]: This should be moved to other module... I'm leaving it here for the time being to see if you like the implementation (which requires the `buffer` index)

**Comment by psaiz** [comment]
I'm opening a new PR from this one. It will require changes in three modules

**Pull request closed without merging**